File: TP2/java/Pokemon1.java
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.stream.Collectors;

public class Pokemon1 implements Cloneable {
    int id;
    int generation;
    int captureRate;
    String name;
    String description;
    ArrayList<String> types;
    ArrayList<String> abilities;
    double weight;
    double height;
    boolean isLegendary;
    Date captureDate;

    public Pokemon1() {
        this(0, 0, 0, "", "", new ArrayList<>(), new ArrayList<>(), 0.0, 0.0, false, null);
    }

    public Pokemon1(int id, int generation, int captureRate, String name, String description, ArrayList<String> types,
            ArrayList<String> abilities, double weight, double height, boolean isLegendary, Date captureDate) {
        this.id = id;
        this.generation = generation;
        this.captureRate = captureRate;
        this.name = name;
        this.description = description;
        this.types = types;
        this.abilities = abilities;
        this.weight = weight;
        this.height = height;
        this.isLegendary = isLegendary;
        this.captureDate = captureDate;
    }

    // Getters
    public int getId() { return id; }
    public int getGeneration() { return generation; }
    public int getCaptureRate() { return captureRate; }
    public String getName() { return name; }
    public String getDescription() { return description; }
    public ArrayList<String> getTypes() { return types; }
    public ArrayList<String> getAbilities() { return abilities; }
    public double getWeight() { return weight; }
    public double getHeight() { return height; }
    public boolean isLegendary() { return isLegendary; }
    public Date getCaptureDate() { return captureDate; }

    @Override
    public Pokemon1 clone() {
        Pokemon1 cloned = null;
        try {
            cloned = (Pokemon1) super.clone();
            cloned.types = new ArrayList<>(this.types);
            cloned.abilities = new ArrayList<>(this.abilities);
            if (this.captureDate != null) {
                cloned.captureDate = (Date) this.captureDate.clone();
            }
        } catch (CloneNotSupportedException e) {
            throw new AssertionError();
        }
        return cloned;
    }

    // Função para imprimir os Pokémons de acordo com o ID 
    public void imprimir() {
        SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy");
        String captureDateStr = (captureDate != null) ? sdf.format(captureDate) : "null";

        // Formatando types e abilities
        String formattedTypes = types.stream()
                .map(type -> "'" + type + "'")
                .collect(Collectors.joining(", "));
        String formattedAbilities = abilities.stream()
                .map(ability -> "'" + ability + "'")
                .collect(Collectors.joining(", "));

        System.out.println("[#" + id + " -> " + name + ": " + description + " - [" + formattedTypes + "] - ["
                + formattedAbilities + "] - " + weight + "kg - " + height + "m - " + captureRate + "% - "
                + isLegendary + " - " + generation + " gen] - " + captureDateStr);
    }

    public static String[] parseCSVLine(String line) {
        List<String> tokens = new ArrayList<>();
        StringBuilder sb = new StringBuilder();
        boolean inQuotes = false;
        int i = 0;
        while (i < line.length()) {
            char c = line.charAt(i);
            if (c == '"') {
                inQuotes = !inQuotes; // Alterna o estado
            } else if (c == ',' && !inQuotes) {
                tokens.add(sb.toString());
                sb = new StringBuilder();
            } else {
                sb.append(c);
            }
            i++;
        }
        tokens.add(sb.toString());
        return tokens.toArray(new String[0]);
    }

    public static boolean isNumeric(String str) {
        boolean result = false;
        try {
            Integer.parseInt(str.trim());
            result = true;
        } catch (NumberFormatException e) {
            // resultado permanece false
        }
        return result;
    }

    public static boolean isDouble(String str) {
        boolean result = false;
        try {
            Double.parseDouble(str.trim());
            result = true;
        } catch (NumberFormatException e) {
            // resultado permanece false
        }
        return result;
    }

    public static boolean isValidDate(String str, SimpleDateFormat dateFormat) {
        boolean result = false;
        try {
            dateFormat.parse(str.trim());
            result = true;
        } catch (Exception e) {
            // resultado permanece false
        }
        return result;
    }

    public static Pokemon1 buscarPokemonPorId(int id, List<Pokemon1> pokemons) {
        Pokemon1 foundPokemon = null;
        int i = 0;
        while (foundPokemon == null && i < pokemons.size()) {
            Pokemon1 p = pokemons.get(i);
            if (p.getId() == id) {
                foundPokemon = p;
            }
            i++;
        }
        return foundPokemon;
    }

    // Função para buscar Pokémon por nome 
    public static boolean buscarPokemonPorNome(String nome, List<Pokemon1> pokemons, Comparacao comparacao) {
        for (Pokemon1 p : pokemons) {
            comparacao.increment();
            if (p.getName().equalsIgnoreCase(nome)) {
                return true;
            }
        }
        return false;
    }

    // Classe para contar comparações
    static class Comparacao {
        private int contador;

        public Comparacao() {
            this.contador = 0;
        }

        public void increment() {
            this.contador++;
        }

        public int getContador() {
            return contador;
        }
    }

    public static void main(String[] args) {
        List<Pokemon1> pokemons = new ArrayList<>();

        // Leitura de todos os Pokémons do arquivo /tmp/pokemon.csv
        try {
            File file = new File("/tmp/pokemon.csv");


            Scanner sc = new Scanner(file);

            // Ignorar a primeira linha (cabeçalho)
            if (sc.hasNextLine()) {
                sc.nextLine();
            }

            SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy");
            dateFormat.setLenient(false);

            while (sc.hasNextLine()) {
                String linha = sc.nextLine();
                if (!linha.isEmpty()) {
                    String[] dados = parseCSVLine(linha);
                    if (dados.length >= 12) {
                        int id = isNumeric(dados[0]) ? Integer.parseInt(dados[0].trim()) : 0;
                        int generation = isNumeric(dados[1].trim()) ? Integer.parseInt(dados[1].trim()) : 0;
                        String name = dados[2].trim();
                        String description = dados[3].trim();

                        String typesStr = dados[4].trim();
                        typesStr = typesStr.replaceAll("[\\[\\]'\"]", "");
                        List<String> types = Arrays.asList(typesStr.split(",\\s*"));

                        boolean isLegendary = Boolean.parseBoolean(dados[5].trim());

                        String abilitiesStr = dados[6].trim();
                        abilitiesStr = abilitiesStr.replaceAll("[\\[\\]'\"]", "");
                        List<String> abilities = Arrays.asList(abilitiesStr.split(",\\s*"));

                        double weight = isDouble(dados[7].trim()) ? Double.parseDouble(dados[7].trim()) : 0.0;
                        double height = isDouble(dados[8].trim()) ? Double.parseDouble(dados[8].trim()) : 0.0;
                        int captureRate = isNumeric(dados[9].trim()) ? Integer.parseInt(dados[9].trim()) : 0;

                        String dateStr = dados[10].trim();
                        Date captureDate = isValidDate(dateStr, dateFormat) ? dateFormat.parse(dateStr) : null;

                        // Adicionar o Pokémon à lista
                        pokemons.add(new Pokemon1(id, generation, captureRate, name, description,
                                new ArrayList<>(types), new ArrayList<>(abilities), weight, height,
                                isLegendary, captureDate));
                    }
                }
            }

            sc.close();
        } catch (Exception e) {
            e.printStackTrace();
        }

        // Agora, ler os IDs da entrada padrão e construir a lista de Pokémons selecionados
        List<Pokemon1> selectedPokemons = new ArrayList<>();

        Scanner inputScanner = new Scanner(System.in);
        String inputLine = inputScanner.nextLine().trim();

        while (!inputLine.equals("FIM")) {
            if (!inputLine.isEmpty()) {
                try {
                    int id = Integer.parseInt(inputLine);
                    Pokemon1 p = buscarPokemonPorId(id, pokemons);
                    if (p != null) {
                        selectedPokemons.add(p.clone()); // Adicionar uma cópia para evitar problemas de aliasing
                    }
                } catch (NumberFormatException e) {
                    // Ignorar entradas que não sejam números
                }
            }

            if (inputScanner.hasNextLine()) {
                inputLine = inputScanner.nextLine().trim();
            } else {
                break;
            }
        }

        // Leitura dos nomes a serem pesquisados
        List<String> nomesPesquisados = new ArrayList<>();

        if (inputScanner.hasNextLine()) {
            inputLine = inputScanner.nextLine().trim();
            while (!inputLine.equals("FIM")) {
                if (!inputLine.isEmpty()) {
                    nomesPesquisados.add(inputLine);
                }

                if (inputScanner.hasNextLine()) {
                    inputLine = inputScanner.nextLine().trim();
                } else {
                    break;
                }
            }
        }

        inputScanner.close();

        // Início da medição de tempo
        long inicio = System.currentTimeMillis();
        Comparacao comparacao = new Comparacao();

        // Realiza as pesquisas sequenciais
        for (String nomePesquisado : nomesPesquisados) {
            boolean encontrado = buscarPokemonPorNome(nomePesquisado, selectedPokemons, comparacao);
            if (encontrado) {
                System.out.println("SIM");
            } else {
                System.out.println("NAO");
            }
        }

        // Fim da medição de tempo
        long fimTempo = System.currentTimeMillis();
        long tempoExecucao = fimTempo - inicio;

        String matricula = "819886"; 
        String nomeArquivo = matricula + "_sequencial.txt";

        try (FileWriter fw = new FileWriter(nomeArquivo)) {
            fw.write(matricula + "\t" + tempoExecucao + "\t" + comparacao.getContador() + "\n");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: TP2/c/Pokemon1.c
//pokemon pesquisa binaria

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>
#include <ctype.h>

#define MAX_POKEMONS 1000
#define MAX_LINE_LENGTH 1024
#define MAX_NAME_LENGTH 100
#define MAX_TYPES_ABILITIES 10

typedef struct {
    int id;
    int generation;
    int captureRate;
    char name[MAX_NAME_LENGTH];
    char description[MAX_LINE_LENGTH];
    char types[MAX_TYPES_ABILITIES][MAX_NAME_LENGTH];
    int typesCount;
    char abilities[MAX_TYPES_ABILITIES][MAX_NAME_LENGTH];
    int abilitiesCount;
    double weight;
    double height;
    bool isLegendary;
    char captureDate[20];
} Pokemon;

typedef struct {
    int contador;
} Comparacao;

void incrementarComparacao(Comparacao *comp) {
    comp->contador++;
}

void replaceChar(char *str, char find, char replace) {
    char *currentPos = strchr(str, find);
    while (currentPos) {
        *currentPos = replace;
        currentPos = strchr(currentPos + 1, find);
    }
}

void removeChars(char *str, const char *charsToRemove) {
    char *src, *dst;
    for (src = dst = str; *src != '\0'; src++) {
        const char *c = charsToRemove;
        bool remove = false;
        while (*c != '\0') {
            if (*src == *c) {
                remove = true;
                break;
            }
            c++;
        }
        if (!remove) {
            *dst++ = *src;
        }
    }
    *dst = '\0';
}

bool parseCSVLine(char *line, Pokemon *pokemon) {
    char *token;
    int fieldCount = 0;
    bool inQuotes = false;
    for (int i = 0; line[i]; i++) {
        if (line[i] == '"') {
            inQuotes = !inQuotes;
        }
        if (line[i] == ',' && inQuotes) {
            line[i] = ';';
        }
    }

    char *saveptr1;
    token = strtok_r(line, ",", &saveptr1);
    while (token != NULL) {
        replaceChar(token, ';', ',');

        switch (fieldCount) {
            case 0:
                pokemon->id = atoi(token);
                break;
            case 1:
                pokemon->generation = atoi(token);
                break;
            case 2:
                strncpy(pokemon->name, token, MAX_NAME_LENGTH - 1);
                pokemon->name[MAX_NAME_LENGTH - 1] = '\0';
                break;
            case 3:
                strncpy(pokemon->description, token, MAX_LINE_LENGTH - 1);
                pokemon->description[MAX_LINE_LENGTH - 1] = '\0';
                break;
            case 4: {
                removeChars(token, "[]'\"");
                char *typeToken;
                char *saveptr2;
                typeToken = strtok_r(token, ",", &saveptr2);
                pokemon->typesCount = 0;
                while (typeToken != NULL && pokemon->typesCount < MAX_TYPES_ABILITIES) {
                    strncpy(pokemon->types[pokemon->typesCount], typeToken, MAX_NAME_LENGTH - 1);
                    pokemon->types[pokemon->typesCount][MAX_NAME_LENGTH - 1] = '\0';
                    pokemon->typesCount++;
                    typeToken = strtok_r(NULL, ",", &saveptr2);
                }
                break;
            }
            case 5:
                pokemon->isLegendary = (strcmp(token, "True") == 0 || strcmp(token, "true") == 0);
                break;
            case 6: {
                removeChars(token, "[]'\"");
                char *abilityToken;
                char *saveptr2;
                abilityToken = strtok_r(token, ",", &saveptr2);
                pokemon->abilitiesCount = 0;
                while (abilityToken != NULL && pokemon->abilitiesCount < MAX_TYPES_ABILITIES) {
                    strncpy(pokemon->abilities[pokemon->abilitiesCount], abilityToken, MAX_NAME_LENGTH - 1);
                    pokemon->abilities[pokemon->abilitiesCount][MAX_NAME_LENGTH - 1] = '\0';
                    pokemon->abilitiesCount++;
                    abilityToken = strtok_r(NULL, ",", &saveptr2);
                }
                break;
            }
            case 7:
                pokemon->weight = atof(token);
                break;
            case 8:
                pokemon->height = atof(token);
                break;
            case 9:
                pokemon->captureRate = atoi(token);
                break;
            case 10:
                strncpy(pokemon->captureDate, token, sizeof(pokemon->captureDate) - 1);
                pokemon->captureDate[sizeof(pokemon->captureDate) - 1] = '\0';
                break;
            default:
                break;
        }
        fieldCount++;
        token = strtok_r(NULL, ",", &saveptr1);
    }

    return fieldCount >= 11;
}

Pokemon *buscarPokemonPorId(int id, Pokemon *pokemons, int totalPokemons) {
    for (int i = 0; i < totalPokemons; i++) {
        if (pokemons[i].id == id) {
            return &pokemons[i];
        }
    }
    return NULL;
}

int compararPokemons(const void *a, const void *b) {
    Pokemon *p1 = (Pokemon *)a;
    Pokemon *p2 = (Pokemon *)b;
    return strcasecmp(p1->name, p2->name);
}

bool buscarPokemonPorNomeBinaria(char *nome, Pokemon *pokemons, int totalPokemons, Comparacao *comparacao) {
    int left = 0;
    int right = totalPokemons - 1;

    while (left <= right) {
        int mid = (left + right) / 2;
        incrementarComparacao(comparacao);
        int cmp = strcasecmp(nome, pokemons[mid].name);
        if (cmp == 0) {
            return true;
        } else if (cmp < 0) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    return false;
}

int main() {
    Pokemon pokemons[MAX_POKEMONS];
    int totalPokemons = 0;
    const char *filePath = "/tmp/pokemon.csv";

    FILE *file = fopen(filePath, "r");
    if (file == NULL) {
        printf("Erro ao abrir o arquivo %s\n", filePath);
        return 1;
    }

    char line[MAX_LINE_LENGTH];
    fgets(line, sizeof(line), file);

    while (fgets(line, sizeof(line), file)) {
        if (totalPokemons >= MAX_POKEMONS) {
            printf("Limite de Pokémons atingido.\n");
            break;
        }
        line[strcspn(line, "\r\n")] = '\0';
        if (parseCSVLine(line, &pokemons[totalPokemons])) {
            totalPokemons++;
        }
    }
    fclose(file);

    Pokemon selectedPokemons[MAX_POKEMONS];
    int totalSelected = 0;

    char inputLine[MAX_LINE_LENGTH];
    while (fgets(inputLine, sizeof(inputLine), stdin)) {
        inputLine[strcspn(inputLine, "\r\n")] = '\0';
        if (strcmp(inputLine, "FIM") == 0) {
            break;
        }
        int id = atoi(inputLine);
        Pokemon *p = buscarPokemonPorId(id, pokemons, totalPokemons);
        if (p != NULL && totalSelected < MAX_POKEMONS) {
            selectedPokemons[totalSelected++] = *p;
        }
    }

    qsort(selectedPokemons, totalSelected, sizeof(Pokemon), compararPokemons);

    char nomesPesquisados[MAX_POKEMONS][MAX_NAME_LENGTH];
    int totalNomes = 0;

    while (fgets(inputLine, sizeof(inputLine), stdin)) {
        inputLine[strcspn(inputLine, "\r\n")] = '\0';
        if (strcmp(inputLine, "FIM") == 0) {
            break;
        }
        strncpy(nomesPesquisados[totalNomes], inputLine, MAX_NAME_LENGTH - 1);
        nomesPesquisados[totalNomes][MAX_NAME_LENGTH - 1] = '\0';
        totalNomes++;
    }

    clock_t inicio = clock();
    Comparacao comparacao = {0};

    for (int i = 0; i < totalNomes; i++) {
        bool encontrado = buscarPokemonPorNomeBinaria(nomesPesquisados[i], selectedPokemons, totalSelected, &comparacao);
        if (encontrado) {
            printf("SIM\n");
        } else {
            printf("NAO\n");
        }
    }

    clock_t fim = clock();
    double tempoExecucao = ((double)(fim - inicio)) / CLOCKS_PER_SEC * 1000;

    char matricula[] = "819886";
    char nomeArquivo[50];
    sprintf(nomeArquivo, "%s_binaria.txt", matricula);

    FILE *logFile = fopen(nomeArquivo, "w");
    if (logFile != NULL) {
        fprintf(logFile, "%s\t%.2lf\t%d\n", matricula, tempoExecucao, comparacao.contador);
        fclose(logFile);
    } else {
        printf("Erro ao criar o arquivo de log.\n");
    }

    return 0;
}
